### PR TITLE
[RPC] Fix decoderawtransaction decoding of segwit txs

### PR DESCRIPTION
**Pull request by achow101**
Makes decoderawtransaction check for the marker and flag bytes present in segwit transactions and decodes them as such in order to avoid situations where segwit transactions are decoded as non-segwit transactions.

As per IRC discussion: https://botbot.me/freenode/bitcoin-core-dev/2017-01-12/?msg=79232896&page=1

Edit: Instead of checking for the bytes, it will just decode the transactions with the segwit formatting as that is the same behavior.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
fundrawtransaction.py fails on travis

**Comment by achow101** [comment]
It is failing because the transaction is a 0 input transaction which is being interpreted as a segwit transaction. As per the discussion on IRC, not being able to decode 0 input transactions is acceptable behavior.

**Comment by sdaftuar** [comment]
I know I suggested on IRC that we just let decoderawtransaction() fail on transactions with no inputs, but perhaps not working at all on 0-input transactions is a bigger regression than necessary?

I believe we only have ambiguity in the case where something might deserialize into a 0-input, 1-output transaction when witness serialization is disabled.  So perhaps we should just assume the segwit deserialization if a transaction can be decoded both ways (not perfect, but the best we can do for now, I think), and otherwise allow for 0-input transactions?

Specifically: I think we can leave the existing call to DecodeHexTx unchanged (ie, pass in `true` as the last argument), and then add an additional check that, if mtx is a 0-input transaction, try to decode again with that last boolean set to `false` (to force segwit deserialization), and if it succeeds, prefer the segwit deserialization over the 0-input one.

We should document this behavior as well...

**Comment by achow101** [comment]
I suppose that would work as well. It wouldn't actually be checking for just 0 inputs but rather the `0x0001` flag that signals for segwit, so not all 0-input txs would have to go through both desers.

**Comment by achow101** [comment]
I rebased and made @sdaftuar 's suggestion. I can squash if this will be merged as this may not yet be the best solution.

Also, it still fails fundrawtransaction.py, at this test: https://github.com/bitcoin/bitcoin/blob/master/qa/rpc-tests/fundrawtransaction.py#L588. I don't know what to do about that.

**Comment by luke-jr** [comment]
What does this actually fix? `DecodeHexTx` *already* tries both deserializations, and only accepts them if the entire data is consumed...

**Comment by jnewbery** [comment]
I've tested this and it doesn't appear to solve the originally reported problem in https://bitcointalk.org/index.php?topic=1748353.msg17477509#msg17477509 :

```
bitcoin-cli decoderawtransaction "02000000000101c564a62f94c025ac80137817d8658aabceaaad30412facecba1bd2255182e1c40000000000ffffffff01ca124c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b7024730440220679eaf5e41eee49b38f3112ec90b49a655db21677db4d7fa80de67aeb987161102204024b85730fc106d0b870623fc221946e624413363eb627fe90eb5047d35565c012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd5600000000"
{
  "txid": "54ad9b3b24064c4033814ddc712a393cfa4e011eb7ac24374687511cd056eac5",
  "hash": "54ad9b3b24064c4033814ddc712a393cfa4e011eb7ac24374687511cd056eac5",
  "size": 191,
  "vsize": 191,
  "version": 2,
  "locktime": 0,
  "vin": [
  ],
  "vout": [
    {
      "value": 27203371073.07775233,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_LEFT 7817d8658aabceaaad30412facecba1bd22551 OP_SIZE OP_UNKNOWN OP_UNKNOWN 0 0 0 0 0 OP_INVALIDOPCODE OP_INVALIDOPCODE OP_INVALIDOPCODE OP_INVALIDOPCODE -74 4c000000000016001443aac20a116e09ea4f OP_PICK be1c55e4c17aa600b7024730440220679eaf5e41 OP_UNKNOWN OP_UNKNOWN OP_BOOLOR f3112ec90b49a655db21677db4d7fa80de67aeb987161102204024b85730fc106d0b870623fc221946e624413363eb627fe90eb5047d3556 12 33 3428659 OP_UNKNOWN [error]",
        "hex": "80137817d8658aabceaaad30412facecba1bd2255182e1c40000000000ffffffff01ca124c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b7024730440220679eaf5e41eee49b38f3112ec90b49a655db21677db4d7fa80de67aeb987161102204024b85730fc106d0b870623fc221946e624413363eb627fe90eb5047d35565c012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd56",
        "type": "nonstandard"
      }
    }
  ]
}
```

am I doing something wrong?

**Comment by achow101** [comment]
You definitely did something wrong. It worked fine for me:

```￼
decoderawtransaction 02000000000101c564a62f94c025ac80137817d8658aabceaaad30412facecba1bd2255182e1c40000000000ffffffff01ca124c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b7024730440220679eaf5e41eee49b38f3112ec90b49a655db21677db4d7fa80de67aeb987161102204024b85730fc106d0b870623fc221946e624413363eb627fe90eb5047d35565c012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd5600000000
￼
{
  "txid": "a24cec50d5cf861d1af4b634f8ed1968c0e9484724bfef5af7f8c383605978c8",
  "hash": "54ad9b3b24064c4033814ddc712a393cfa4e011eb7ac24374687511cd056eac5",
  "size": 191,
  "vsize": 110,
  "version": 2,
  "locktime": 0,
  "vin": [
    {
      "txid": "c4e1825125d21bbaecac2f4130adaaceab8a65d817781380ac25c0942fa664c5",
      "vout": 0,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "30440220679eaf5e41eee49b38f3112ec90b49a655db21677db4d7fa80de67aeb987161102204024b85730fc106d0b870623fc221946e624413363eb627fe90eb5047d35565c01", 
        "03335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd56"
      ],
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 0.04985546,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 43aac20a116e09ea4f7914be1c55e4c17aa600b7",
        "hex": "001443aac20a116e09ea4f7914be1c55e4c17aa600b7",
        "type": "witness_v0_keyhash"
      }
    }
  ]
}
```

**Comment by sipa** [comment]
I don't understand what this PR is doing. As @luke-jr points out, `DecodeHexTx` always tries both deserializations (in fact, everything always tries both)... but it prefers the non-witness one when `fTryNoWitness` is passed, which is the case. If we want it to prefer the witness version in case of ambiguity, we should just pass false.

**Comment by sipa** [comment]
Oh, I see... you're looking at the hex code in order to determine the number of txins. Can you instead look at tx.vin.size() of the decode?

**Comment by achow101** [comment]
We discussed this on IRC a while ago. The issue is with 0-input transactions, specifically with witness transactions being interpreted as non-withess 0 input transactions. This PR is to make it prefer the witness serialization over a non-witness for 0-input transactions but maintain the original behavior for other transactions.

**Comment by sipa** [comment]
@achow101 Isn't that just identical to passing `false` to DecodeHexTx? The only case in which the decoding is ambiguous is where the non-witness transaction has 0 inputs.

**Comment by achow101** [comment]
@sipa I don't think so. As I understand it, when you pass `false` to `DecodeHexTx` it will interpret a non-witness 0-input transaction as a witness transaction and thus fail to decode it. What this does is that such transactions will be decoded properly on the first pass with `true` being passed in so as to still be able to decode non-witness 0-input transactions instead of failing with decoding them as witness transactions.

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -521,10 +525,16 @@ UniValue decoderawtransaction(const JSONRPCRequest& request)
...
     RPCTypeCheck(request.params, boost::assign::list_of(UniValue::VSTR));
 
-    CMutableTransaction mtx;
+    CMutableTransaction mtx, witMtx;
 
-    if (!DecodeHexTx(mtx, request.params[0].get_str(), true))
+    if (!DecodeHexTx(mtx, request.params[0].get_str(), true)) {
         throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+    }
+    if (request.params[0].get_str().compare(8, 4, "0001") == 0) {
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: I think this can be written as `if (mtx.vin.size() == 0) {` instead.

**Comment by sipa** [comment]
@achow101 I see, of course. The deserialization code doesn't actually try two combinations.

**Comment by achow101** [comment]
I'm baffled as to why the test magically pass now with @sipa's fix.

**Comment by luke-jr** [comment]
It's no longer checking `vout.size()==1`? (which it probably ought to..)

**Comment by sipa** [comment]
A transaction with 0 inputs and 2 outputs can be mistaken for an (invalid) extended transaction as well.

**Comment by achow101** [comment]
> A transaction with 0 inputs and 2 outputs can be mistaken for an (invalid) extended transaction as well.

Right, but that shouldn't be a problem now because it will properly decode as non-witness, fail to decode as witness and thus fallback to non-witness decoding.

**Comment by jnewbery** [comment]
Apologies. I had an old bitcoind in my path. I've retested and this now works.

I can explain why fundrawtransaction passes. It's nothing to do with the change in decoderawtransaction. You've rebased off master and picked up #9650, which causes deserialization to fail if there are junk bytes at the end of a transaction. The test was failing on:

`decoderawtransaction "0100000000010000000000000000066a047465737400000000"`

With this PR and without #9650, we'd:

- deserialize with fTryNoWitness set to True, and succeed in deserializing a non-witness 0-in,1-out transaction
- deserialize with fTryNoWitness set to False, and succeed in deserializing a witness 0-in,0-out transaction and ignoring the trailing junk bytes
- return the witness 0-in,0-out transaction

With #9650 we:

- deserialize with fTryNoWitness set to True, and succeed in deserializing a non-witness 0-in,1-out transaction
- deserialize with fTryNoWitness set to False, and fail because of the trailing junk bytes
- return the witness 0-in,1-out transaction

(see - no magic :smile: )

A couple of general comments:

- Can you add a code comment above the new code in `decoderawtransaction()` explaining why we're deserializing two ways? It's quite subtle, so some explanation here would be good.
- Can you add the same fix to bitcoin-tx (src/bitcoin-tx.cpp)? It's a pain, but while we have two tools for signing/decoding transactions, we should try to keep them aligned.

**Comment by achow101** [comment]
@jnewbery Ah, that makes sense now.

> Can you add a code comment above the new code in decoderawtransaction() explaining why we're deserializing two ways? It's quite subtle, so some explanation here would be good.

Sure.

> Can you add the same fix to bitcoin-tx (src/bitcoin-tx.cpp)? It's a pain, but while we have two tools for signing/decoding transactions, we should try to keep them aligned.

I assume this should go in around the only line that has `DecodeHexTx()`?

**Comment by jnewbery** [comment]
> I assume this should go in around the only line that has DecodeHexTx()?

I think it needs a bit more thought than that. bitcoin-tx combines the functionality of createrawtransaction, signrawtransaction and decoderawtransaction. With your change here, the deserialization logic is different for those different RPCs. Should we try to mirror that in bitcoin-tx?

**Comment by luke-jr** [comment]
@achow101 This doesn't even build now...

**Comment by achow101** [comment]
whoops. Bad copy-paste. Fixed that and rebased onto master.

**Comment by luke-jr** [comment]
Better if you could rebase on 9828f9a9962c1bee5c343847030b9cfd87a40a5e

**Comment by achow101** [comment]
It would be nice (perhaps necessary?) to have this in by the time segwit activates.

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -521,10 +525,21 @@ UniValue decoderawtransaction(const JSONRPCRequest& request)
...
 
-    CMutableTransaction mtx;
+    CMutableTransaction mtx, witMtx;
 
-    if (!DecodeHexTx(mtx, request.params[0].get_str(), true))
+    // First attempt to decode the transaction as a non-witness transaction
+    if (!DecodeHexTx(mtx, request.params[0].get_str(), true)) {
         throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+    }
+    // If the transaction has 0-inputs (either a 0-input non-witness transaction or a normal transaction)
```
Comment (jnewbery) [inline src/rpc/rawtransaction.cpp]: what do you mean by 'normal transaction' here?

**Comment by jnewbery** [comment]
I don't like that this breaks parity between `bitcoin-tx sign` and `bitcoin-cli signrawtransaction`.

`bitcoin-tx sign` will now decode the transaction as segwit when there's ambiguity, but `bitcoin-cli signrawtransaction` will decode the transaction as 0-input non segwit when there's ambiguity.

File: src/rpc/rawtransaction.cpp (hunk 2 above)
Comment (achow101) [inline src/rpc/rawtransaction.cpp]: that should be "witness transaction"

**Comment by achow101** [comment]
@jnewbery I can remove the commit that changes bitcoin-tx

**Comment by jnewbery** [comment]
> I can remove the commit that changes bitcoin-tx

But then you've broken the parity between `bitcoin-tx` and `bitcoin-cli decoderawtransaction`

**Comment by achow101** [comment]
@jnewbery I'm not quite sure what you are asking me to do. Do you want to have `decoderawtransaction`, `signrawtransaction`, and `bitcoin-tx` all have the same decoding behavior?

**Comment by jnewbery** [comment]
I'd suggest making bitcoin-tx behave like signrawtransaction if it's being used to sign a transaction, and behave like createrawtransaction/decoderawtransaction if it's being used to decode, create or modify (without signing) a transaction.

**Comment by achow101** [comment]
@jnewbery I don't think it is possible to have separate behavior like that in bitcoin-tx. You can have multiple things happening in one bitcoin-tx command. You can have it modify a transaction and then sign it, and for all of those to act on the same transaction consistently, the transaction needs to be the same across all bitcoin-tx commands, so there can't be separate decoding behavior. Also, the transaction is only decoded once for all bitcoin-tx commands.

**Pull request closed without merging**